libflux/src/semantic/parser/mod.rs

```diff
@@ -0,0 +1,1729 @@
+    parser.parse_polytype()
+}
+
+// Parse passes the source text through the Lexer, receives tokens and removes
+// whitespace. Its then initializes parsing and returns a PolyType AST.
+// This is the only function meant to be accessed by the end user. It handles 
+// both lexing and parsing of string polytypes. 
+pub fn Parse(source: &str) -> Result<PolyType, &'static str> {
+    let tokens = Lex(source); 
+    let mut clean_tokens: Vec<Token> = Vec::new(); 
+
+    for token in tokens {
+        if token.token_type != TokenTypes::WHITESPACE {
+            clean_tokens.push(token);
+        }
+    }
+
+    let mut parser = Parser{ tokens: clean_tokens.iter().peekable() }; 
+
+    parser.parse_polytype()
+}
+
+impl Parser<'_> {
+    // next grabs the next token using the Iter()'s next method and unpacks
+    // the value if there are still tokens to parse
+    fn next(& mut self) -> Token {
+        match self.tokens.next() {
+            Some(token) => (*token).clone(),
+            None => Token {
+                token_type: TokenTypes::EOF,
+                text: None
+            },
+        }
+    }
+    
+    // peek returns a preview of the next Token using Iter()'s peek method
+    // and unpacks the value if there are still tokens to parse
+    fn peek(& mut self) -> Token {
+        match self.tokens.peek() {
+            Some(token) => (**token).clone(),
+            None => Token {
+                token_type: TokenTypes::EOF,
+                text: None
+            },
+        }
+    }
+
+    // parse_polytype steps through the token list and checks that each 
+    // token is in the correct order. 
+
+    // Error handling for Parser's methods needs to be improved so that more 
+    // helpful messages are returned when parsing fails. 
+    fn parse_polytype(& mut self) -> Result<PolyType, &'static str> {
+        if self.next().token_type != TokenTypes::FORALL {
+            return Err("Missing forall")
+        }
+        if self.next().token_type != TokenTypes::LEFTSQUAREBRAC {
+            return Err("Missing left square bracket")
+        }
+
+        let free_vars = self.parse_vars();
+        
+        let ok_free_vars = match free_vars {
+            Err(e) => { return Err(e) }
+            Ok(free_vars) => free_vars
+        };
+
+        if self.next().token_type != TokenTypes::RIGHTSQUAREBRAC {
+            return Err("Missing right square bracket")
+        }
+
+        let mut cons = None; 
+        if self.next().token_type == TokenTypes::WHERE {
+            cons = self.parse_constraints();
+        }
+
+        let expr = self.parse_monotype();
+
+        match expr {
+            Ok(expr) => {
+                Ok(PolyType {
+                free: ok_free_vars,
+                cons: cons,
+                expr: expr,
+                })
+            }
+            Err(e) => Err(e)
+        }
+        
+    }
+
+
+    /**Production rules mentioned below can be found in the accompanying grammar.md file */
```

Comment:
Do all of the `parse_*` names match the names of the production rules in the grammar? If not can you make them match? I don't have a preference of whether we change the functions or the grammar, choose whatever name you think is best.

Reply:
Yeah, that's a reasonable request. It also eliminates the need for the repetitive comments on  the various parsing methods. I think I will keep the grammar as is and change the functions.

Reply:
Ended up making some small changes to both.